### ISSUE TITLE
Error while running on windows

### ISSUE THREAD

**Issue by craouette** [obuchtala/swig#11]
Hi,

sorry to use the issues to contact you, but I didn't find an other way. I compiled swig-v8 with CMake and VCC on windows. Everything seems to worked. I also updated the installed Lib/Javascript/v8 directory manually from the Lib folder. I set up the SWIG_LIB env. variable. 
But when I run the command:
swig -c++ -javascript -
v8  ....\Examples\javascript\class\example.i
The program crash.
If I run the same command with python, it work fine!

any idea? Am I missing some steps in the installation?

Thanks for the help,

Pierre

**Comment by obuchtala** [comment]
Hi Pierre,

actually, the javascript module lacks totally testing under windows.

ATM we are beginning to work on making this branch SWIG main tree ready.
Part of that work will also be to make it running under windows.

So for now, you have to be patient...
Or even better: you could lend a hand and making it work under windows.

E.g., a first step would be to find out why the current version crashes 
under windows.

Cheers,
Oliver

**Comment by craouette** [comment]
Is it possible to get the parse.c file? just in order to be sure the one I generated is ok!!

Edit:
I think it is ok... but it is a mess to debug... swig hides all types behind (void *)!!! difficult to know if the type is a char \* or a DohBase \* ... 
The error is in int V8Emitter::enterClass(Node *n)
when calling Template t_decl_class(getTemplate("jsv8_declare_class_template"));
the returned value is not the one created .
The method Template JSEmitter::getTemplate(const String *name) creates the templates t on the stack... so it is freed when the function exists. Or Am I missing something?

**Comment by craouette** [comment]
Any reaction on my comment?

**Comment by obuchtala** [comment]
Hmmm... yes debugging is rather difficult due to the voids...

I can't tell right now from that what is wrong there.
I will most certainly give it a try during this week.

I let you know when it is basically working...

Cheers,
Oliver

**Comment by craouette** [comment]
I think the problem is in this piece of code (from javascript.cxx, line 646):
Template JSEmitter::getTemplate(const String *name) {
  String *templ = Getattr(templates, name);

  if (!templ) {
    Printf(stderr, "Could not find template %s\n.", name);
    SWIG_exit(EXIT_FAILURE);
  }

  Template t(templ, name);

  return t;
}

The object t is stack allocated, which means it will be invalid when leaving the function getTemplate.
refer to for a better explanation: http://www.informit.com/articles/article.aspx?p=25033&seqNum=3

It is exactly what is happening on window. t is valid before leaving the method, and invalid when arriving in the calling function/method.

I may be wrong though...

**Comment by obuchtala** [comment]
Nope that is not the problem, that is "return by value".

Hmm.. I think I need to take a look myself. Can't do that right now as I don't have a proper
window setup running. However, I planned to do this during this week or the next anyways.

Thank you for your efforts. I let you know when I have a bit more of a glue.

**Comment by craouette** [comment]
Yes, but the pointers (String *code; and   String *templateName;)  are just copied to the copy. then the original is freed before leaving the method Template JSEmitter::getTemplate(const String *name).
which means that the 2 pointers are pointing to invalid memory!
This is because there is a destructor but not a copy constructor.

when in the method, I have:
-       (DohBase *)(t.code) 0x01509e90 {data=0x01403da0 type=0x005e6808 meta=0x00000000 ...}
-       (DohBase *)(t.templateName) 0x01509ea0 {data=0x01403e28 type=0x005e6808 meta=0x00000000 ...}

but the result in the calling code is:
-       (DohBase *)(t_decl_class.code)  0x01509ea0 {data=0x00000000 type=0x00000000 meta=0x00000000 ...}
-       (DohBase *) (t_decl_class.templateName) 0x01509e90 {data=0x01509ea0 type=0x00000000 meta=0x00000000 ...}  

and latter, the empty type cause the crash.

**Comment by obuchtala** [comment]
Ok.

With GCC it works with the assignment operator...
Could you implement the copy ctor and try if that fixes the problem?

**Comment by craouette** [comment]
yep, I will.

**Comment by obuchtala** [comment]
Thanx :)

**Comment by craouette** [comment]
Done, and it is fixing the problem... At least I am getting a example_wrap.cxx file which seems correct... Need to compile it to be sure!

**Comment by obuchtala** [comment]
Great! Thanks :)

Probably the generated code won't work out-of-the-box. We are having those version incompatibilities with v8
and your version does not yet take care of it properly.
See #12 for details.
